File: packages/executable-stories-pytest/src/executable_stories/_json_writer.py
"""Simple JSON serialization for RawRun output."""

from __future__ import annotations

import json
import os
import tempfile
from typing import Any


def write_raw_run(raw_run: dict[str, Any], output_path: str) -> None:
    """Write a RawRun dict to a JSON file.

    Creates parent directories if they don't exist. The file appears whole:
    it is written alongside the destination and then renamed over it, so a
    reader never sees a half-written run and a crash mid-write leaves the
    previous file intact.
    """
    parent = os.path.dirname(output_path) or "."
    os.makedirs(parent, exist_ok=True)

    fd, temp_path = tempfile.mkstemp(dir=parent, prefix=".raw-run-", suffix=".json")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            json.dump(raw_run, f, indent=2)
            f.write("\n")
        os.replace(temp_path, output_path)
    except BaseException:
        if os.path.exists(temp_path):
            os.unlink(temp_path)
        raise

File: packages/executable-stories-ruby/lib/executable_stories/json_writer.rb
# frozen_string_literal: true

require "json"
require "fileutils"
require_relative "types"

module ExecutableStories
  module JsonWriter
    module_function

    def write_raw_run(run, output_path)
      dir = File.dirname(output_path)
      FileUtils.mkdir_p(dir) unless Dir.exist?(dir)

      json = JSON.pretty_generate(ExecutableStories.run_to_h(run)) + "\n"
      File.write(output_path, json)
    end
  end
end